Wip rgw d4n with write cache

mosayyebzadeh opened the pull request:
This pull request adds write cache part into the latest D4N code.

samarahu committed:
RGW: Add cache driver base class

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/d4n: This commit squashes the following commits related to
d4n policy and filter driver.

RGW: Add cache driver base class - add d4n filter driver.
RGW: Update cache and policy files; add RedisDriver into D4N filter
RGW: Add D4N classes and unit testing; update cpp_redis submodule
QA: Add D4N teuthology suite
RGW: Improve D4N readability and structure
RGW: Add base D4N policy and test
RGW: Add GWF policy to D4N
RGW: Add global weight property
RGW: Added D4N namespace
RGW: Update policy driver interface
RGW: Update unit tests
RGW: Add Address struct to all files
RGW: Update D4N names and structure
RGW: Update structure and fix errors
RGW: Add more features to policy driver
RGW: Access local weight in policy code
RGW: Work on D4N workflows
RGW: Add object directory class and remove copy_value method
RGW: Establish object directory in read workflow
RGW: Update cache and policy files; add RedisDriver into D4N filter
RGW: Switch out D4N cache methods with Redis driver methods
RGW: Update RedisDriver to match new CacheDriver structure; define set_attrs method
RGW: Fix D4N read workflow crashes
RGW: Update D4N files to match CacheDriver changes
RGW: fix key_exists method for RedisDriver and clean up rgw_sal_d4n.cc
RGW: Use correct get_block method
RGW: Make CachePolicy a virtual class
RGW: Initialize localWeight if not found and develop find_victim method
RGW: Debugging network failure
RGW: Rebase and debugging network failure
RGW: Update RedisDriver::list_entries and usage in D4N policy driver
RGW: Fix network failure issue; add entries and entry methods
RGW: Update D4N test files to match rebase
RGW: Update D4N policy and RedisDriver with entries
RGW: make localWeight an attribute
rgw/cache: commit to fix compilation issues.

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/cache: This commit squashes modifications to base class.

1. Rename namespace from rgw::cal to rgw::cache
2. Remove Partition from the CacheDriver constructor.
3. Removing rgw_cache_driver.cc since all methods
are pure virtual in rgw_cache_driver.h.
4. Moving structures Partition and Entry out of
class CacheDriver. And modifying the signatures of methods
get_current_partition_info() and get_free_space().
5. modifying the cache driver base class.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: this commit squashes the following commits for ssd backed cache driver.

initial implementation of ssd backed cache driver.This contains implementations
for methods related to Entry and Partition.

adding methods for put, get and delete data.
They are all synchronous.

added CacheAioRequest to handle async read and write
requests and added a corresponding read op in aio.h and aio.cc.

added get_async method implementation and
initial implementation to handle async reads using asio.

changing the signature of list_entries to match that with
base class and making it public.

adding implementation for asynchronous get.

modifying namespace from rgw::cal to rgw::cache.
Modifying the SSDDriver constructor for Partition related
modifications.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/cache: This commit squashes the following commits related to list_entries cache api.

RGW: Add list_entries method to CacheDriver and to Redis driver.
RGW: Change CacheDriver::list_entries output type
RGW: Update RedisDriver::list_entries and usage in D4N policy driver

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
RGW: add entries and entry methods

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
RGW: Fix find_client method; rgw crash issue

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
RGW: Fix rgw crash issue

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
RGW: move CacheDriver instance to D4N Filter

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
RGW: Add D4N chunking mechanism

Signed-off-by: Samarah Uriarte <samarah.uriarte@ibm.com>

samarahu committed:
rgw/cache: This commit squashes commits related to boost::redis library.

RGW: Add boost_redis library
RGW: Add boost_redis to CMake
RGW: Add aedis methods to RedisDriver
RGW: Minor RedisDriver cleanup
build: Remove build/boost_redis
build: remove src/redis
build: Set src/boost_redis submodule to release version
RGW: Remove broken boost_redis
RGW: Add boost_redis
rgw: Get rgw_redis_driver to build
rgw/redis: Update `boost::redis` library
rgw/cache: fixing compilation error, by removing aedis header
file and src file includes from redis driver

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Co-authored-by: Adam Emerson <aemerson@redhat.com>
Co-authored-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/cache: This commit adds implementation of Partition in
ssd driver and squashes the following related commits.

RGW: Minor RedisDriver cleanup
RGW: Add partitioning
RGW: Cleanup RedisDriver

Signed-off-by: Samarah Uriarte <samarah.uriarte@ibm.com>

samarahu committed:
RGW: Work on oid in cache and CacheBlock block ID

Signed-off-by: Samarah Uriarte <samarah.uriarte@ibm.com>

adamemerson committed:
build: Remove build/boost_redis

Signed-off-by: Adam Emerson <aemerson@redhat.com>

pritha-srivastava committed:
rgw/cache: This commit squashes the following commits
related to xattrs in ssd driver.

rgw/cache: code to handle xattrs, which are stored as
file xattrs.
rgw/cache: correcting ssd driver get_attrs.
rgw/cache: fixing set_attrs in ssd cache driver.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
RGW: Update read workflow so attributes get written

Signed-off-by: Samarah Uriarte <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/cache: implementation of put_async method in ssd driver
and redis driver and squashes the following commits.

rgw/cache: implementation of async put. the call
does not take into account throttling for now.
rgw/cache: dummy implementation of put_async in redis
driver to fix compilation error.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/d4n: completing the read flow for a local read cache,
with ssd cache driver as the backend.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
test/d4n: this commit squashes the following commits related to tests for d4n.

test/redis: Add temporary `boost::redis` unit test program for experimentation
test/rgw: Add/update unit test programs for D4N classes + RedisDriver; switch to `boost::redis` usage
test/d4n: Begin `cpp_redis` removal from D4N filter test program; needs
to be fixed
test/d4n: Temporarily remove D4N filter test program until D4N workflows
are fully implemented

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/cache: this commit squashes following commits related to changes in
redis driver and ssd driver.

rgw/redis: Add `del`, `put_async`, and `calculate_free_space` methods to RedisDriver
rgw/redis: Work on RedisDriver `put_async` and `get_async` methods
rgw/cache: Add error-checking to `SSDDriver::set_attr`
rgw/redis: Remove `calculate_free_space` method from RedisDriver, fix
`free_space` calculations, clean up async code

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/cache: this commit adds `optional_yield` to ssd and redis driver
and squashes the following commits.

RGW: Add `optional_yield` parameter to Cache Driver methods
RGW: Add `optional_yield` parameter to SSD Driver methods
rgw/cache: fix compilation errors related to optional_yield.

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Co-authored-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: fixing the segmentation fault by passing in 'yield'
by value to async_completion.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: this commit squashes commits related to removing
cache_aio_read()/cache_aio_write() from cache api, ssd driver
and redis driver.

rgw/cache: removing cache_aio_read()/cache_aio_write()
interfaces from the Cache API. And moving cache_read_op
from rgw_aio.cc to ssd cache driver.
rgw/cache: changes related to cache_aio_read()/cache_aio_write()
in redis cache driver.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: prototype implementation of an LRU Cache which includes
indexing and a list.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: Modified APIs after removing indexing from
cache driver implementation.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: this commit removes indexing from ssd and redis
driver and squashes the following commits.

rgw/cache: modifications to the ssd cache driver after removing
indexing from the cache backend.
rgw/cache: modifications to redis cache driver, after
removing indexing from cache backend.
rgw/cache: modifications to policy driver code, commenting out
some pieces which refer to cache apis. Those apis have been
removed as part of modification to remove indexing from the
cache backend. These need to appropriately re-instated, once
we finalize which module will contain the data structures
related to cache policy.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: moving LRU related data structures and interfaces to the Policy Manager.
Replacing the related calls in d4n filter driver with that of the cachepolicy in the
Policy Manager and adding cache block to the block directory
once the block is written to the cache backend and the
policy manager data structure.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: fixing free space calcuation when data is deleted
from the cache backend.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: This commit fixes two things in ssd backed cache backend:
1. fixing free space calcuation when data is deleted.

2. implementation of append_data().

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw: Change D4N and RedisDriver `ifndef` directives to `#pragma once`; make integer errors negative; fix filter includes

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/policy: this commit squashes the following changes to policy and filter driver files.

d4n/policy: Update policy with correct directory usage; create LFUDAEntry struct + entries map; add `optional_yield` to `insert` method
d4n/policy: Add `version` to Entry struct/policy methods and calls in
the D4N filter; add error checking and fix logic in LFUDA code

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/directory: Update directory `exist_key` method to take in
block/object; fix endpoint generation; add `remove_host` method

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/d4n: Remove `cacheBlock` and commented out read workflow from D4N filter

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/d4n: this commit squashes the following commits related to
clean up and removal of cpp_redis.

d4n/filter: Add `optional_yield` to CacheDriver calls in D4N Filter and
make minor updates to several filter methods
rgw/d4n: fix compilation issue.
rgw/cache: Add `del` method to CacheDriver and SSDDriver
cmake/d4n: Remove unnecessary D4N lines
rgw: Add `io_context` to D4N Filter and RedisDriver, remove `cpp_redis`
library from RedisDriver, and perform minor cleanup
d4n: Remove `cpp_redis` library from D4N directory and policy; update calls in filter; move Entry struct to base CachePolicy class
build/cpp_redis: Remove `cpp_redis` library
rgw/d4n: including <boost/asio/detached.hpp> wheerever needed.
rgw/d4n : fixes to d4n filter, policy, directory and redis driver files
for compilation errors.

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Signed-off-by: Casey Bodley <cbodley@redhat.com>
Co-authored-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw: Add `rgw_local_cache_address` yaml config

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/cache: saving dpp in set_client_cb() instead
of flush_write_part() so that the correct value
is available everywhere in handle_data().

Saving dpp value in flush_last_part() caused a crash in
rgw in handle_data().

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: adding d4n config options for ssd backed
cache driver.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
d4n: Add connection client names for policy and directory

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/policy: Update policy to use block name

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n: Store `optional_yield` in client cb and update index used in
D4N/cache classes

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: copying over read op params of the filter driver to
the read op params of 'next' in prepare() method, to be used
while storing objects to the backend store.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: this commit squashes the following two commits.

rgw/d4n: Remove `get_block` from policy and add logic to D4N filter
d4n/policy: Create a min heap of the local weights

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/filter: Change from copying block to object in `copy_obj`

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/cache: fixing update_attrs and get_attrs in ssd backed cache backend.

update_attrs did not check whether an xattr already existed before
modifying it - added check for that. corrected converting bufferlist
to string.

get_attrs did not check whether attribute size is zero or negative -
added check for that.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
rgw/d4n: this commit squashes the following commits related to LFUDA Policy.

rgw/d4n: Make minor updates, improve `LFUDAPolicy::erase` method, and
fix `localWeight` policy sets
rgw/d4n: Update `LFUDA::find_victim` method, add locks in policy code,
and remove setting of attrs in filter's `put_async` calls
d4n/policy: Update logic for minimum average weight calculation
d4n/policy: Remove iteration in `eviction`

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/cache: fixing the free space available in ssd backed cache
backend by taking into account the size of xattrs in addition to
the file size.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: adding unittests for ssd backed cache backend, and
fixing bugs that the tests revealed.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
d4n/directory: Add `ObjectDirectory::update_field` method

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/cache: this commit adds entries to redis driver and squashes the
following two commits.

rgw: Update D4N policy and RedisDriver with entries
rgw/d4n: fixing compilation error in d4n policy code.

Signed-off-by: Samarah <samarah.uriarte@ibm.com>
Co-authored-by: Pritha Srivastava <prsrivas@redhat.com>

adamemerson committed:
rgw: Thread site config into RADOS driver

Make this available so multisite sync doesn't have to rely on the zone
service.

Signed-off-by: Adam Emerson <aemerson@redhat.com>

samarahu committed:
d4n: Update block `mtime` in filter and change `creationTime` to string
in directory due to seg faults

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: restructuring LRUPolicy::erase(), update() and eviction() method.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
d4n/policy: Fix minor logical errors

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: using 'id_tag' as 'version' of an object for non-versioned
objects. Using 'instance' of 'oid' as 'version' for
versioned objects.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

samarahu committed:
d4n/policy: Initialize `age` to 1

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/cache: Make protected members private

Signed-off-by: samarah <suriarte@redhat.com>

samarahu committed:
rgw: Take in `io_context` by reference

Signed-off-by: samarah <suriarte@redhat.com>

samarahu committed:
d4n/policy: Make `shutdown` method specific to LFUDA

Signed-off-by: samarah <suriarte@redhat.com>

samarahu committed:
rgw/cache: Make `bl` and `attrs` params `const` for cache write ops

Signed-off-by: samarah <suriarte@redhat.com>

pritha-srivastava committed:
rgw/cache: removing interface related to Partition.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: removing `experimental` fallback from ssd backed cache driver.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: changing `get_attr` api such that it returns
int signifying success or error, and `attr_val` is added
as an out param.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

pritha-srivastava committed:
rgw/cache: this commit squashes commits for removing CephContext* from
ssd, redis driver and related test files.

rgw/cache: removing CephContext* from ssd driver and redis driver.
It can be accessed in every api using dpp->get_cct(), where dpp
is of type DoutPrefixProvider*.
rgw/test: Remove `cct` parameter from cache driver initialization call
redis driver test file.
rgw/qa: removing cct parameter from ssd driver test file.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>
Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/filter: Fix cache `exist_key` check to ensure the return value is positive

Signed-off-by: samarah <suriarte@redhat.com>

samarahu committed:
d4n/policy: Fix `find_victim` logic so victim is found in `eviction`
method

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/policy: Add additional lock guards for other `entries_heap` accesses

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/filter: Update `creationTime` calculation

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/filter: Correct `CacheObj` `hostsList` value so it's empty since the
object isn't being locally cached

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/cache: This commit squashes the following commits for redis driver.

RGW: fix key_exists method for RedisDriver and clean up rgw_sal_d4n.cc
RGW: Implement RedisDriver::get_free_space
rgw/cache: modifying namespace from rgw::cal to rgw::cache.
RGW: Update D4N files to match CacheDriver changes
RGW: Fix D4N read workflow crashes
RGW: Update RedisDriver to match new CacheDriver structure; define set_attrs method
RGW: Switch out D4N cache methods with Redis driver methods
RGW: Update Cache Driver structure
RGW: Update cache files.
RGW: create redis cache driver files

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw: this commit squashes commits performing cleanup in various files
like d4n filter, policy, cache backend, aio and process_env.

d4n/filter: Remove unnecessary erasure of attrs in read op's `prepare`
method
d4n: Remove `dout` definitions from header files
rgw: Remove unnecessary includes in `rgw_process_env.h`
rgw: Remove `cache_read_op` from `rgw_aio.h`
rgw: Return POSIX error codes and perform minor cleanup for D4N and
Redis files

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
test/redis: Remove temporary boost.redis test program

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n/policy: Change `size_t` argument type and add error checking to
`set_local_weight_sum` method

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
d4n: Add a central Redis connection to the filter to be shared by the rest of the D4N classes and update D4N unit tests appropriately

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

samarahu committed:
rgw/d4n: Perform minor cleanup; add error checking for `localWeights` and fix order of entry deletion in policy

Signed-off-by: Samarah <samarah.uriarte@ibm.com>

pritha-srivastava committed:
rgw/d4n: modifications to the filter driver:
1. replaced put_async() with put() in handle_data()
2. moved calls to update() from iterate() to flush()

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

cbodley committed:
rgw/d4n: fixing the 'get' call hanging issues and
Operation Canceled errors, by wrapping async_exec
call in boost::asio::dispatch() call.

Signed-off-by: Casey Bodley <cbodley@redhat.com>

pritha-srivastava committed:
rgw: updating the boost_redis submodule version to point to
`develop` branch.

Signed-off-by: Pritha Srivastava <prsrivas@redhat.com>

mosayyebzadeh committed:
combining write cache with latest D4N code

mosayyebzadeh committed:
cleaning the code

mosayyebzadeh committed:
removing some bugs on bigger objects

samarahu commented on src/common/options/rgw.yaml.in:
```diff
@@ -3926,6 +3936,16 @@ options:
+- name: rgw_d4n_cache_cleaning_interval
+  type: int
+  level: advanced
+  desc: The rgw directory port
```
This description doesn't seem correct

mosayyebzadeh replied:
Noted!

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -894,27 +902,141 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+            }
+	  }
+	} else{
+            ldpp_dout(save_dpp, 1) << "D4NFilterObject::D4NFilterWrtieOp::process" << __func__ << "(): ERROR: wrtiting data to the backend failed!" << dendl;
```
Minor typo: D4NFilterWrtieOp --> D4NFilterWriteOp

samarahu commented on src/rgw/rgw_cache_driver.h:
```diff
@@ -25,6 +25,7 @@ class CacheDriver {
     virtual int put_async(const DoutPrefixProvider* dpp, const std::string& key, const bufferlist& bl, uint64_t len, const rgw::sal::Attrs& attrs) = 0;
     virtual int append_data(const DoutPrefixProvider* dpp, const::std::string& key, const bufferlist& bl_data, optional_yield y) = 0;
     virtual int delete_data(const DoutPrefixProvider* dpp, const::std::string& key, optional_yield y) = 0;
+    virtual int rename(const DoutPrefixProvider* dpp, const::std::string& oldKey, const::std::string& newKey, optional_yield y) = 0;
```
The RedisDriver needs an implementation for this method as well; I can help with that

samarahu commented:
The unit test programs for the classes that were changed (LFUDAPolicy, SSDDriver, etc) need to be updated, as well. I can help with these updates.

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -395,6 +409,175 @@ bool LFUDAPolicy::erase(const DoutPrefixProvider* dpp, const std::string& key, o
+    	  std::string oid_in_cache = "D_" + prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_lst);  	  
+    	  std::string new_oid_in_cache = prefix + "_" + std::to_string(fst) + "_" + std::to_string(cur_lst);
+      	  ldpp_dout(dpp, 20) << __func__  << ": " << __LINE__ << ": oid_in_cache is: " << oid_in_cache << dendl;
+    	  cacheDriver->get(dpp, oid_in_cache, fst, cur_lst, data, obj_attrs, null_yield);
```
@mosayyebzadeh : please use a valid `yield_context` instead of `null_yield`.

mosayyebzadeh replied:
I am not sure how to get a valid yield_context here.

pritha-srivastava replied:
ok, yes we do not have access to the yield_context in the `init` method too, for the time being null yield is ok, we can come back to this later also.

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -395,6 +409,175 @@ bool LFUDAPolicy::erase(const DoutPrefixProvider* dpp, const std::string& key, o
+      	    return;
+    	  }
+
+    	  cacheDriver->rename(dpp, oid_in_cache, new_oid_in_cache, null_yield);
```
are you renaming the dirty blocks to remove the prefix 'D' once they are written to the storage backend? Wasn't it decided that blocks/objects would either exist in the cache or in the storage backend? and hence delete them from the cache backend, once they are written back to the storage backend.

mosayyebzadeh replied:
no, the delete action should be done through LFUDA policy. There is a high probability that the written data will be read soon. an example is ML workloads.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -894,27 +907,139 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+    	dirty = false;
+	
+  	ldpp_dout(save_dpp, 10) << "D4NFilterObject::D4NFilterWriteOp::" << __func__ << "(): calling next process" << dendl;
+	ret = next->process(std::move(data), offset);
```
is next->process writing to the storage backend? if `d4n_writecache` is false, why is the policy data structure being updated here? I understand that directory needs to be updated to for dirty/ non-dirty blocks. Policy data structure entries track blocks/ objects that are stored in the cache backend.

mosayyebzadeh replied:
Good point. It should be removed.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -894,27 +907,139 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
+
+    int ret = 0;
+
+      if (d4n_writecache == false){
```
correct the indentation.

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -332,7 +332,7 @@ int LFUDAPolicy::eviction(const DoutPrefixProvider* dpp, uint64_t size, optional
 }
 
-void LFUDAPolicy::update(const DoutPrefixProvider* dpp, std::string& key, uint64_t offset, uint64_t len, std::string version, optional_yield y)
+void LFUDAPolicy::update(const DoutPrefixProvider* dpp, std::string& key, uint64_t offset, uint64_t len, std::string version, bool dirty, time_t creationTime, const rgw_user user, optional_yield y)
```
where are these additional params being used?

mosayyebzadeh replied:
We discussed this with Samarah. They are not needed and should be removed.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -894,27 +907,139 @@ int D4NFilterObject::D4NFilterDeleteOp::delete_obj(const DoutPrefixProvider* dpp
-  }*/
+    auto version = obj->get_instance();
+
+    std::string prefix;
```
this versioning will have to be fixed when caching head objects is also in place.

pritha-srivastava commented on src/rgw/driver/d4n/d4n_policy.cc:
```diff
@@ -395,6 +409,175 @@ bool LFUDAPolicy::erase(const DoutPrefixProvider* dpp, const std::string& key, o
+
+    	  cacheDriver->rename(dpp, oid_in_cache, new_oid_in_cache, null_yield);
+
+          ldpp_dout(dpp, 20) << "AMIN: " << __func__ << __LINE__ << dendl;
```
please modify the comments.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -508,9 +508,17 @@ int D4NFilterObject::D4NFilterReadOp::flush(const DoutPrefixProvider* dpp, rgw::
       std::string oid_in_cache = prefix + "_" + std::to_string(ofs) + "_" + std::to_string(len);
       ldpp_dout(dpp, 20) << "D4NFilterObject::" << __func__ << " calling update for offset: " << offset << " adjusted offset: " << ofs  << " length: " << len << " oid_in_cache: " << oid_in_cache << dendl;
-      source->driver->get_policy_driver()->get_cache_policy()->update(dpp, oid_in_cache, ofs, len, version, y);
+      source->driver->get_policy_driver()->get_cache_policy()->update(dpp, oid_in_cache, ofs, len, version, dirty, creationTime,  source->get_bucket()->get_owner(), y);
```
in iterate() method, is reading of dirty blocks not handled/ allowed?

mosayyebzadeh replied:
In the next commit, I have updated the code to reflect this. I had missed these changes through the merge process.

mosayyebzadeh committed:
updating iterate function to check the dirty flag

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -306,106 +375,706 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+              existing_block.blockID = block.blockID;
+              existing_block.size = block.size;
+  	      existing_block.dirty = block.dirty;
+              if (blockDir->get(&existing_block, y) < 0) {
```
Wouldn't setting the `existing_block` data in the above lines (1010, 1011, 1012) and then calling `get` for that same block overwrite those values? This applies on lines 1053-1055 as well.

mosayyebzadeh replied:
We need blockID and size to get the info from the directory. dirty should be removed.

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -738,12 +768,15 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
         block.blockID = ofs;
         block.size = bl.length();
         block.version = version;
+        block.dirty = false; //writing to the backend
```
You can update the comment here - that data is being read from the backend, hence the block is not dirty

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -738,12 +768,15 @@ int D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::handle_data(bufferlist& bl
           if (ret == 0) {
+  	    std::string objEtag = "";
  	    filter->get_policy_driver()->get_cache_policy()->update(dpp, oid, ofs, bl.length(), version, dirty, creationTime,  source->get_bucket()->get_owner(), *y);
+	    filter->get_policy_driver()->get_cache_policy()->updateObj(dpp, prefix, version, dirty, source->get_obj_size(), creationTime, source->get_bucket()->get_owner(), objEtag, *y);
```
what is the purpose of adding a call to `updateObj` here?

mosayyebzadeh replied:
This is the last block of the Object. We want to keep track of all objects in the cache. Although we use o_map_entries in the cleaning function, I think having a complete list of objects in the in-memory structure is a good idea, clean or dirty.

pritha-srivastava replied:
ok

pritha-srivastava commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -602,9 +602,25 @@ int D4NFilterObject::D4NFilterReadOp::iterate(const DoutPrefixProvider* dpp, int
     ldpp_dout(dpp, 20) << "D4NFilterObject::iterate:: " << __func__ << "(): READ FROM CACHE: oid=" << oid_in_cache << " length to read is: " << len_to_read << " part num: " << start_part_num << 
     " read_ofs: " << read_ofs << " part len: " << part_len << dendl;
 
+    std::string key = oid_in_cache;
```
In `flush` method above, for `dirty` objects that start with a prefix `D` you do not pass in the oid that has a prefix `D` and you also do not set the `dirty` flag to true - hence in `update` method, the  `localWeight` will not be updated correctly, because you do that based on adding a prefix `D` after checking for `dirty` flag.

Instead of basing your code by adding a prefix `D`, why don't you add only dirty objects to `o_entries_map`, and then in the cleaning code, get those blocks based on `o_entries_map` and dirty flag and write them to the cache - this way you don't need to have special handling for oid with prefix 'D', and the rename operation also won't be needed.

mosayyebzadeh replied:
We need D_  in case of server failure, so we can recover the dirty data from the local SSD. Otherwise, when the server fails, we will miss all the information about dirty objects.

mosayyebzadeh replied:
flush operation is invoked in the read process, hence the data is clean, and hence dirty flag shouldn't be set to true.

pritha-srivastava replied:
flush() method is used to flush the aio list which comprises of the blocks read from the cache, and since the write workflow has blocks that are dirty, in the read workflow you have already added logic for reading blocks with prefix 'D', the async completion of ios for those blocks in the flush() method needs to be handled correctly.

mosayyebzadeh replied:
We need to discuss this, I am not sure that I am getting your point correctly.

mosayyebzadeh committed:
Updating write cache based on the pull request comments.

Read process needs to be updated based on write process. It needs to check where is the data and if it is dirty or clean.
If it is in the cache and dirty, we need to put D_ in the oid of the object before reading it from cache.
If it is clean, there is nothing to do.

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -306,106 +375,694 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+              }
+	    }
+          } else {
+            ldpp_dout(save_dpp, 1) << "D4NFilterObject::D4NFilterWriteOp::process" << __func__ << "(): ERROR: wrtiting data to the cache failed!" << dendl;
```
Small typo: `ERROR: wrtiting data to the cache failed!` --> `ERROR: writing data to the cache failed!` (also on line 1019)

samarahu commented on src/rgw/driver/d4n/rgw_sal_d4n.cc:
```diff
@@ -306,106 +375,694 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+		    if (blockDir->set(&block, *y) < 0)
+		      ldpp_dout(dpp, 10) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory set method failed." << dendl;
+		  } else {
+		  if (blockDir->update_field(&block, "blockHosts", blockDir->cct->_conf->rgw_local_cache_address, *y) < 0)
+		    ldpp_dout(dpp, 10) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory update_field method failed." << dendl;
```
Please fix indentation

samarahu commented:
Please rebase the branch with the updated `wip-rgw-d4n` branch when you can. Currently, this PR is tracking changes that are old and unrelated to the write cache.

pritha-srivastava commented:
@mosayyebzadeh : The `D4NFilterWriter` will work in case of small objects. In case of larger objects, I believe writer related to Multipart needs to be implemented. Please check with Daniel and Casey about this in the next d4n meeting.

mosayyebzadeh committed:
updating flush functions and comments.

pritha-srivastava commented:
@mosayyebzadeh : As noted in the d4n meeting, the `eviction` method does not handle not evicting dirty objects as of now. Also have you tested this PR in any basic scenario? If done then please update the PR with the steps that you have followed.

pritha-srivastava commented:
@mosayyebzadeh : if a cache has different versions of the same object, how is the cleaning process handling that?
